File: 프로그래머스/lv2/81302. 거리두기 확인하기/거리두기 확인하기.py
from collections import deque

dx=[0,0,1,-1]
dy=[1,-1,0,0]

def bfs(case):
    vis=[[-1]*5 for _ in range(5)]
    for i in range(5):
        for j in range(5):
            if case[i][j]=="P":
                q = deque([(i,j)])
                vis[i][j]=0
                while q:
                    x,y=q.popleft()
                    for dir in range(4):
                        nx=x+dx[dir]
                        ny=y+dy[dir]
                        if nx<0 or ny<0 or nx>=5 or ny>=5:
                            continue
                        if vis[nx][ny]==0:
                            continue
                        if case[nx][ny]=="X":
                            continue
                        if case[nx][ny]=="P":
                            return 0
                        vis[nx][ny]=vis[x][y]+1
                        if vis[nx][ny]==2:
                            continue
                        q.append((nx,ny))
    return 1
def solution(places):
    answer = []
    for place in places:
        answer.append(bfs(place))
    return answer

File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.java
import java.util.*;

class Solution {
    
    private int size = 5;
    private int[] dx = {0, 0, 1, -1};
    private int[] dy = {1, -1, 0, 0};
    
    public int[] solution(String[][] places) {
        int[] answer = new int[size];
        
        for(int i=0; i<size; i++){
            char[][] place = new char[size][size];
            for(int j=0; j<size; j++){
                place[j] = places[i][j].toCharArray();
            }
            
            if(isPossible(place)){
                answer[i] = 1;
            }
        }
        return answer;
    }
    
    public boolean isPossible(char[][] place){
        for(int j=0; j<size; j++){
            for(int k=0; k<size; k++){
                if(place[j][k]=='P'){
                    boolean[][] vis = new boolean[size][size];
                    vis[j][k] = true;
                    if(!dfs(place, vis, j, k, 0)){
                        return false;
                    }
                }
            }
        }
        return true;
    }
    
    public boolean dfs(char[][] place, boolean[][] vis, int x, int y, int dist){
        if(dist==2){
            return true;
        }
        
        boolean check = true;
        
        for(int dir=0; dir<4; dir++){
            int nx = x+dx[dir];
            int ny = y+dy[dir];
            if(nx<0 || nx>=size || ny<0 || ny>=size) continue;
            if(vis[nx][ny]) continue;
            if(place[nx][ny]=='X') continue;
            if(place[nx][ny]=='P') return false;
            vis[nx][ny] = true;
            check = dfs(place, vis, nx, ny, dist+1);
            if(check==false){
                return false;
            }
        }
        
        return check;
    }
}